src/lib/Err.hs

```diff
@@ -42,61 +54,142 @@ data ErrType = NoErr
              | ZipErr
              | EscapedNameErr
              | ModuleImportErr
+             | MonadFailErr
                deriving (Show, Eq)
 
-type Except = Either Err
+type SrcPosCtx  = Maybe SrcPos
+type SrcTextCtx = Maybe (Int, String) -- Int is the offset in the source file
+data ErrCtx = ErrCtx
+  { srcTextCtx :: SrcTextCtx
+  , srcPosCtx  :: SrcPosCtx
+  , messageCtx :: [String] }
+    deriving (Show, Eq)
+
+type Except = Either Errs
 
 type SrcPos = (Int, Int)
-type SrcCtx = Maybe SrcPos
 
-type MonadErr = MonadError Err
+class MonadFail m => Fallible m where
+  throwErrs :: Errs -> m a
+  addErrCtx :: ErrCtx -> m a -> m a
+
+-- We have this in its own class because IO and `Either Err` can't implement it
```

Comment:
This is not clear, i feel like it is missing a
"But XYZ *can* and should implement it."

Reply:
I think it's good that we have this comment, because you can't see what's _not_ implemented from the code, but you can grep for `instance.*CtxReader XYZ` if you want. So IMO pretty good after all.